File: project/jimmer-apt/src/main/java/org/babyfish/jimmer/apt/transactional/TxGenerator.java
package org.babyfish.jimmer.apt.transactional;

import com.squareup.javapoet.*;
import org.babyfish.jimmer.apt.Context;
import org.babyfish.jimmer.apt.GeneratorException;
import org.babyfish.jimmer.apt.MetaException;
import org.babyfish.jimmer.apt.immutable.generator.Annotations;
import org.babyfish.jimmer.apt.immutable.generator.Constants;

import javax.lang.model.element.*;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import java.io.IOException;

public class TxGenerator {

    private final Context ctx;

    private final TypeElement typeElement;

    private final TypeMirror sqlClientType;

    private final TypeMirror runtimeExceptionType;

    private final String sqlClientName;

    private final AnnotationMirror classTx;

    private TypeSpec.Builder typeBuilder;

    public TxGenerator(Context ctx, TypeElement typeElement) {
        this.ctx = ctx;
        this.typeElement = typeElement;
        this.sqlClientType = ctx.getElements().getTypeElement("org.babyfish.jimmer.sql.JSqlClient").asType();
        this.runtimeExceptionType = ctx.getElements().getTypeElement("java.lang.RuntimeException").asType();
        this.sqlClientName = sqlClientName();
        this.classTx = TxUtil.tx(ctx, typeElement);
    }

    private String sqlClientName() {
        VariableElement sqlClientElement = null;
        for (Element element : typeElement.getEnclosedElements()) {
            if (element.getKind() == ElementKind.FIELD) {
                VariableElement variableElement = (VariableElement) element;
                if (!variableElement.getModifiers().contains(Modifier.STATIC)) {
                    if (ctx.getTypes().isAssignable(sqlClientType, variableElement.asType())) {
                        if (sqlClientElement != null) {
                            throw new MetaException(
                                    element,
                                    "The class uses @Tx cannot multiple non-static sqlClient fields"
                            );
                        }
                        sqlClientElement = variableElement;
                    }
                }
            }
        }
        if (sqlClientElement == null) {
            throw new MetaException(
                    typeElement,
                    "The class uses @Tx must have a non-static field whose type is JSqlClient"
            );
        }
        if (sqlClientElement.getModifiers().contains(Modifier.PRIVATE)) {
            throw new MetaException(
                    sqlClientElement,
                    "The sqlClient field of the class uses @Tx cannot be private, default or protected is recommended"
            );
        }
        return sqlClientElement.getSimpleName().toString();
    }

    public void generate() {
        typeBuilder = TypeSpec
                .classBuilder(typeElement.getSimpleName().toString() + "Tx")
                .superclass(typeElement.asType());
        for (AnnotationMirror annotationMirror : typeElement.getAnnotationMirrors()) {
            TypeElement annotationElement = (TypeElement) ctx.getTypes().asElement(annotationMirror.getAnnotationType());
            String qualifiedName = annotationElement.getQualifiedName().toString();
            if (!TxUtil.TARGET_ANNOTATION.equals(qualifiedName) && !TxUtil.TX.equals(qualifiedName)) {
                typeBuilder.addAnnotation(AnnotationSpec.get(annotationMirror));
            }
        }
        AnnotationMirror targetAnnotation = TxUtil.targetAnnotation(ctx, typeElement);
        if (targetAnnotation != null) {
            String qualifiedName = targetAnnotation.getElementValues().values().iterator().next().getValue().toString();
            typeBuilder.addAnnotation(ClassName.bestGuess(qualifiedName));
        }
        if (typeElement.getModifiers().contains(Modifier.PUBLIC)) {
            typeBuilder.addModifiers(Modifier.PUBLIC);
        }
        if (typeElement.getModifiers().contains(Modifier.ABSTRACT)) {
            typeBuilder.addModifiers(Modifier.ABSTRACT);
        }

        addConstructors();
        addMethods();
        try {
            JavaFile
                    .builder(
                            ((PackageElement) typeElement.getEnclosingElement()).getQualifiedName().toString(),
                            typeBuilder.build()
                    )
                    .indent("    ")
                    .build()
                    .writeTo(ctx.getFiler());
        } catch (IOException ex) {
            throw new GeneratorException(
                    String.format(
                            "Cannot generate tx class for '%s'",
                            typeElement.getQualifiedName().toString()
                    ),
                    ex
            );
        }
    }

    private void addConstructors() {
        for (Element element : typeElement.getEnclosedElements()) {
            if (element.getKind() == ElementKind.CONSTRUCTOR && !element.getModifiers().contains(Modifier.PRIVATE)) {
                ExecutableElement executableElement = (ExecutableElement) element;
                MethodSpec.Builder builder = MethodSpec.constructorBuilder();
                StringBuilder sb = new StringBuilder();
                for (VariableElement variableElement : executableElement.getParameters()) {
                    builder.addParameter(ParameterSpec.get(variableElement));
                    if (sb.length() != 0) {
                        sb.append(", ");
                    }
                    sb.append(variableElement.getSimpleName().toString());
                }
                builder.addStatement("super(" + sb + ')');
                typeBuilder.addMethod(builder.build());
            }
        }
    }

    private void addMethods() {
        for (Element element : typeElement.getEnclosedElements()) {
            if (!(element instanceof ExecutableElement)) {
                continue;
            }
            AnnotationMirror tx = TxUtil.tx(ctx, element);
            if (tx != null) {
                if (element.getKind() != ElementKind.METHOD) {
                    throw new MetaException(
                            element,
                            "Member that is not method cannot be decorated by @Tx"
                    );
                }
                if (element.getModifiers().contains(Modifier.STATIC)) {
                    throw new MetaException(
                            element,
                            "Static method cannot be decorated by @Tx"
                    );
                }
                if (element.getModifiers().contains(Modifier.PRIVATE)) {
                    throw new MetaException(
                            element,
                            "Private method cannot be decorated by @Tx"
                    );
                }
                if (element.getModifiers().contains(Modifier.FINAL)) {
                    throw new MetaException(
                            element,
                            "Final method cannot be decorated by @Tx"
                    );
                }
                if (element.getModifiers().contains(Modifier.ABSTRACT)) {
                    throw new MetaException(
                            element,
                            "Abstract method cannot be decorated by @Tx"
                    );
                }
                for (TypeMirror thrownType : ((ExecutableElement)element).getThrownTypes()) {
                    if (!ctx.getTypes().isAssignable(runtimeExceptionType, thrownType)) {
                        TypeElement thrownElement = (TypeElement) ctx.getTypes().asElement(thrownType);
                        throw new MetaException(
                                element,
                                "Method decorated by @Tx can only throw RuntimeException, but it throws \"" +
                                        thrownElement.getQualifiedName() +
                                        "\""
                        );
                    }
                }
            }
            if (tx == null &&
                    classTx != null &&
                    element.getModifiers().contains(Modifier.PUBLIC) &&
                    element.getKind() == ElementKind.METHOD &&
                    !element.getModifiers().contains(Modifier.STATIC) &&
                    !element.getModifiers().contains(Modifier.PRIVATE)) {
                if (element.getModifiers().contains(Modifier.FINAL)) {
                    throw new MetaException(
                            element,
                            "The public method inherits the class-level @Tx cannot be final"
                    );
                }
                tx = classTx;
            }
            if (tx != null) {
                addMethod((ExecutableElement) element, tx);
            }
        }
    }

    private void addMethod(ExecutableElement executableElement, AnnotationMirror tx) {
        MethodSpec.Builder builder = MethodSpec
                .methodBuilder(executableElement.getSimpleName().toString())
                .addAnnotation(Override.class);
        if (executableElement.getModifiers().contains(Modifier.PROTECTED)) {
            builder.addModifiers(Modifier.PROTECTED);
        }
        if (executableElement.getModifiers().contains(Modifier.PUBLIC)) {
            builder.addModifiers(Modifier.PUBLIC);
        }
        String doc = ctx.getElements().getDocComment(executableElement);
        if (doc != null) {
            builder.addJavadoc(doc.replace("$", "$$"));
        }
        for (AnnotationMirror annotationMirror : executableElement.getAnnotationMirrors()) {
            TypeElement annotationElement = (TypeElement) ctx.getTypes().asElement(annotationMirror.getAnnotationType());
            String qualifiedName = annotationElement.getQualifiedName().toString();
            if (!"java.lang.Override".equals(qualifiedName) && !TxUtil.TX.equals(qualifiedName)) {
                builder.addAnnotation(AnnotationSpec.get(annotationMirror));
            }
        }
        for (TypeParameterElement typeParameterElement : executableElement.getTypeParameters()) {
            builder.addTypeVariable(TypeVariableName.get(typeParameterElement));
        }
        builder.returns(TypeName.get(executableElement.getReturnType()));
        StringBuilder sb = new StringBuilder();
        for (VariableElement variableElement : executableElement.getParameters()) {
            builder.addParameter(ParameterSpec.get(variableElement));
            if (sb.length() != 0) {
                sb.append(", ");
            }
            sb.append(variableElement.getSimpleName().toString());
        }
        for (TypeMirror thrownType : executableElement.getThrownTypes()) {
            builder.addException(TypeName.get(thrownType));
        }
        Element propagationElement = Annotations.annotationValue(tx, "value", null);
        String propagation = propagationElement != null ?
                propagationElement.getSimpleName().toString() :
                "REQUIRED";
        boolean isVoid = executableElement.getReturnType().getKind() == TypeKind.VOID;
        CodeBlock.Builder cb = CodeBlock.builder();
        cb.add(
                "$L$L.transaction($T.$L, () -> $> {\n",
                isVoid ? "" : "return ",
                sqlClientName,
                Constants.PROPAGATION_CLASS_NAME,
                propagation
        );
        cb.add(
                "$Lsuper.$L($L);\n",
                isVoid ? "" : "return ",
                executableElement.getSimpleName().toString(),
                sb.toString()
        );
        if (isVoid) {
            cb.add("return null;\n");
        }
        cb.add("$<});");
        builder.addCode(cb.build());
        typeBuilder.addMethod(builder.build());
    }
}

File: project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/transactional/TxGenerator.kt
package org.babyfish.jimmer.ksp.transactional

import com.google.devtools.ksp.*
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.Dependencies
import com.google.devtools.ksp.symbol.*
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ksp.toAnnotationSpec
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import org.babyfish.jimmer.ksp.*
import org.babyfish.jimmer.ksp.immutable.generator.PROPAGATION_CLASS_NAME
import org.babyfish.jimmer.ksp.util.fastResolve
import org.babyfish.jimmer.ksp.util.suppressAllAnnotation
import java.io.OutputStreamWriter

class TxGenerator(
    private val codeGenerator: CodeGenerator,
    private val ctx: Context,
    private val declaration: KSClassDeclaration
) {
    private val simpleName = declaration.simpleName.asString() + "Tx"

    private val sqlClientName: String =
        determineSqlClientName()

    private val classTx = declaration.annotation(TX)

    fun generate(allFiles: List<KSFile>) {
        codeGenerator.createNewFile(
            Dependencies(false, *allFiles.toTypedArray()),
            declaration.packageName.asString(),
            simpleName
        ).use {
            val fileSpec = FileSpec
                .builder(
                    declaration.packageName.asString(),
                    simpleName
                ).apply {
                    indent("    ")
                    addAnnotation(suppressAllAnnotation())
                    addType()
                }.build()
            val writer = OutputStreamWriter(it, Charsets.UTF_8)
            fileSpec.writeTo(writer)
            writer.flush()
        }
    }

    private fun determineSqlClientName(): String {
        val sqlClientType = ctx.resolver.getClassDeclarationByName("org.babyfish.jimmer.sql.kt.KSqlClient")!!.asStarProjectedType()
        val props = declaration.getDeclaredProperties()
            .filter { sqlClientType.isAssignableFrom(it.type.fastResolve()) }
            .toList()
        if (props.isEmpty()) {
            throw MetaException(
                declaration,
                "The class uses @Tx must have a non-static properties whose type is KSqlClient"
            )
        }
        if (props.size > 1) {
            throw MetaException(
                declaration,
                "The class uses @Tx cannot multiple non-static sqlClient properties"
            )
        }
        val prop = props[0]
        if (prop.isPrivate()) {
            throw MetaException(
                prop,
                "The sqlClient field of the class uses @Tx cannot be private, protected or internal is recommended"
            )
        }
        return prop.name
    }

    private fun FileSpec.Builder.addType() {
        addType(
            TypeSpec.classBuilder(simpleName)
                .apply {
                    if (declaration.isInternal()) {
                        addModifiers(KModifier.INTERNAL)
                    }
                    if (declaration.isAbstract()) {
                        addModifiers(KModifier.ABSTRACT)
                    }
                    for (anno in declaration.annotations) {
                        val fullName = anno.fullName
                        if (fullName != TX  && fullName != TARGET_ANNOTATION) {
                            addAnnotation(anno.toAnnotationSpec())
                        }
                    }
                    val targetAnnotation = declaration.annotation(TARGET_ANNOTATION)
                    if (targetAnnotation != null) {
                        val annoDeclaration = targetAnnotation.getClassArgument("value")
                        addAnnotation(annoDeclaration!!.toClassName())
                    }
                    superclass(declaration.toClassName())
                    declaration.primaryConstructor?.let {
                        for (parameter in it.parameters) {
                            addSuperclassConstructorParameter(parameter.name!!.asString())
                        }
                    }
                    addConstructors()
                    addFunctions()
                }
                .build()
        )
    }

    private fun TypeSpec.Builder.addConstructors() {
        val primaryConstructor = declaration.primaryConstructor?.takeIf { !it.isPrivate() }
        if (primaryConstructor !== null) {
            primaryConstructor(
                FunSpec.constructorBuilder().apply {
                    setConstructorProperties(primaryConstructor, true)
                }.build()
            )
        } else {
            for (constructor in declaration.getConstructors()) {
                if (!constructor.isPrivate()) {
                    addFunction(
                        FunSpec.constructorBuilder().apply {
                            setConstructorProperties(constructor, false)
                        }.build()
                    )
                }
            }
        }
    }

    private fun FunSpec.Builder.setConstructorProperties(
        constructor: KSFunctionDeclaration,
        primary: Boolean
    ) {
        if (constructor.isProtected()) {
            addModifiers(KModifier.PROTECTED)
        }
        if (constructor.isInternal()) {
            addModifiers(KModifier.INTERNAL)
        }
        for (anno in constructor.annotations) {
            addAnnotation(anno.toAnnotationSpec())
        }
        for (parameter in constructor.parameters) {
            addParameter(
                ParameterSpec
                    .builder(parameter.name!!.asString(), parameter.type.toTypeName())
                    .build()
            )
        }
        if (!primary) {
            callSuperConstructor(*constructor.parameters.map { it.name!!.asString() }.toTypedArray())
        }
    }

    private fun TypeSpec.Builder.addFunctions() {
        for (function in declaration.getDeclaredFunctions()) {
            val tx = function.annotation(TX)
            if (tx != null && !function.isOpen()) {
                throw MetaException(
                    function,
                    "Only open method cannot be decorated by @Tx"
                )
            }
            val finalTx = when {
                tx !== null -> tx
                classTx === null || !function.isPublic() || function.isConstructor() -> continue
                else -> {
                    if (!function.isOpen()) {
                        throw MetaException(
                            function,
                            "The public method inherits the class-level @Tx must be open"
                        );
                    }
                    classTx
                }
            }
            addFunction(function, finalTx)
        }
    }

    private fun TypeSpec.Builder.addFunction(function: KSFunctionDeclaration, tx: KSAnnotation) {
        val propagation = tx.get<Any>("value").toString().let {
            val index = it.lastIndexOf(".")
            if (index == -1) {
                it
            } else {
                it.substring(index + 1)
            }
        }
        addFunction(
            FunSpec.builder(function.simpleName.asString())
                .apply {
                    addModifiers(KModifier.OVERRIDE)
                    if (function.isProtected()) {
                        addModifiers(KModifier.PROTECTED)
                    } else if (function.isInternal()) {
                        addModifiers(KModifier.INTERNAL)
                    }
                    for (anno in function.annotations) {
                        if (anno.fullName != TX) {
                            addAnnotation(anno.toAnnotationSpec())
                        }
                    }
                    for (parameter in function.parameters) {
                        addParameter(
                            ParameterSpec
                                .builder(parameter.name!!.asString(), parameter.type.toTypeName())
                                .build()
                        )
                    }
                    function.returnType?.let {
                        returns(it.toTypeName())
                    }
                    addCode(
                        CodeBlock.builder().apply {
                            beginControlFlow(
                                "return this.%L.transaction(%T.%L)",
                                sqlClientName,
                                PROPAGATION_CLASS_NAME,
                                propagation
                            )
                            addStatement(
                                "super.%L(%L)",
                                function.simpleName.asString(),
                                function.parameters.map { it.name!!.asString() }.joinToString { ", " }
                            )
                            endControlFlow()
                        }.build()
                    )
                }
                .build()
        )
    }
}